File: design_pattern/solid/srp/solid_srp.ts
import { writeFile } from "fs/promises"

class Journal {
  private entries: string[] = []
  private count: number = 0

  addEntry(text: string): number {
    this.entries.push(`${++this.count}: ${text}`)
    return this.count
  }

  removeEntry(index: number): void {
    this.entries.splice(index, 1)
  }

  toString(): string {
    return this.entries.join("\n")
  }

  getEntries(): string[] {
    return this.entries
  }
}

class Persistence {
  constructor(private readonly lineSeparator: string = "\n") {}

  async saveToFile(journal: Journal, filename: string): Promise<void> {
    const content = journal.getEntries().join(this.lineSeparator)
    await writeFile(filename, content, "utf8")
  }
}

async function main() {
  const journal = new Journal()
  journal.addEntry("I cried today.")
  journal.addEntry("I ate a bug.")
  console.log(journal.toString())

  const persistence = new Persistence("\r\n")
  await persistence.saveToFile(journal, "journal.txt")
}

main().catch(console.error)

File: design_pattern/solid/srp/solid_srp.dart
import 'dart:io';

class Journal {
  final List<String> entries = [];
  int count = 0;

  int addEntry(String text) {
    entries.add('${++count}: $text');
    return count;
  }

  void removeEntry(int index) {
    entries.removeAt(index);
  }

  @override
  String toString() => entries.join('\n');
}

class Persistence {
  final String lineSeparator;

  const Persistence({this.lineSeparator = '\n'});

  Future<void> saveToFile(Journal journal, String filename) async {
    final file = File(filename);
    await file.writeAsString(journal.entries.join(lineSeparator));
  }
}

void main() async {
  final journal = Journal();
  journal.addEntry('I cried today.');
  journal.addEntry('I ate a bug.');
  print(journal.toString());

  final persistence = Persistence(lineSeparator: '\r\n');
  await persistence.saveToFile(journal, 'journal.txt');
}
